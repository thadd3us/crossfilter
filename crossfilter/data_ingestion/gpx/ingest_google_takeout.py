"""CLI program for ingesting Google Takeout Records.json files into parquet."""

import logging
from pathlib import Path

import pandas as pd
import typer
from tqdm.contrib.concurrent import process_map

from crossfilter.core.schema import SchemaColumns as C
from crossfilter.data_ingestion.gpx.google_takeout_parser import load_google_takeout_records_to_df

logger = logging.getLogger(__name__)
typer.main.get_command_name = lambda name: name
app = typer.Typer(pretty_exceptions_show_locals=False)


def find_takeout_files(base_dir: Path) -> list[Path]:
    """Find all Google Takeout Records.json files recursively in the base directory."""
    if not base_dir.exists():
        raise FileNotFoundError(f"Base directory not found: {base_dir}")

    if not base_dir.is_dir():
        raise ValueError(f"Base directory is not a directory: {base_dir}")

    # Look for Records.json files (Google Takeout location history)
    takeout_files = list(base_dir.rglob("Records.json"))
    logger.info(f"Found {len(takeout_files)} Google Takeout Records.json files in {base_dir}")

    return takeout_files


def process_single_takeout_file(takeout_file: Path) -> pd.DataFrame:
    """Process a single Google Takeout Records.json file and return a DataFrame."""
    try:
        return load_google_takeout_records_to_df(takeout_file)
    except Exception as e:
        logger.error(f"Failed to process {takeout_file}: {e}")
        # Return empty DataFrame with correct schema on error
        return pd.DataFrame(
            columns=[
                C.UUID,
                C.DATA_TYPE,
                C.SOURCE_FILE,
                C.TIMESTAMP_MAYBE_TIMEZONE_AWARE,
                C.TIMESTAMP_UTC,
                C.GPS_LATITUDE,
                C.GPS_LONGITUDE,
            ]
        )


@app.command()
def main(
    base_dir: Path = typer.Argument(
        ..., help="Directory to recursively search for Google Takeout Records.json files"
    ),
    output_parquet: Path = typer.Argument(..., help="Parquet file to write to"),
    max_workers: int = typer.Option(None, help="Maximum number of parallel workers"),
) -> None:
    """Ingest Google Takeout Records.json files into parquet format."""
    logging.basicConfig(level=logging.INFO)

    # Find all Takeout files
    takeout_files = find_takeout_files(base_dir)

    if not takeout_files:
        logger.info("No Google Takeout Records.json files found")
        return

    # Process files in parallel
    logger.info(f"Processing {len(takeout_files)} Google Takeout files...")
    dataframes = process_map(
        process_single_takeout_file,
        takeout_files,
        max_workers=max_workers,
        desc="Processing Takeout files",
        chunksize=1,
    )

    # Filter out empty DataFrames
    non_empty_dataframes = [df for df in dataframes if not df.empty]

    if not non_empty_dataframes:
        logger.info("No valid data found in Google Takeout files")
        return

    # Concatenate all DataFrames
    logger.info(f"Concatenating {len(non_empty_dataframes)} DataFrames...")
    combined_df = pd.concat(non_empty_dataframes, ignore_index=True)

    combined_df = combined_df.drop_duplicates()
    combined_df = combined_df.sort_values([C.TIMESTAMP_UTC, C.UUID], ignore_index=True)

    if combined_df[C.UUID].duplicated().any():
        logger.warning(
            "Duplicate UUIDs found in data. This should not happen if the data is clean."
        )
        dupes = combined_df[C.UUID].duplicated(keep=False)
        logger.info(f"Dupes: {combined_df[dupes][C.SOURCE_FILE].value_counts()}")
        raise ValueError("Duplicate UUIDs found in data.")

    logger.info(
        f"Total records: {len(combined_df)} (H3 columns already computed per-file in parallel)"
    )
    combined_df[C.UUID] = combined_df[C.UUID].astype('object')
    combined_df = combined_df.set_index(C.UUID, verify_integrity=True)

    logger.info(f"Writing to parquet file: {output_parquet}")
    combined_df.to_parquet(
        output_parquet,
        engine="pyarrow",
        index=True,
    )

    logger.info("Google Takeout ingestion completed successfully")


if __name__ == "__main__":
    app()