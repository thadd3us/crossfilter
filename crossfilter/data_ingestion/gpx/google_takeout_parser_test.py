"""Tests for Google Takeout location history parsing functions."""

import json
from datetime import datetime, timezone
from pathlib import Path

import pytest
from syrupy.assertion import SnapshotAssertion

from crossfilter.data_ingestion.gpx.google_takeout_parser import (
    convert_e7_coordinate,
    generate_uuid_from_components,
    load_google_takeout_records_to_df,
    parse_location_record,
    parse_takeout_timestamp,
)
from crossfilter.core.schema import DataType


def create_sample_takeout_data() -> dict:
    """Create sample Google Takeout Records.json data for testing."""
    return {
        "locations": [
            {
                "timestamp": "2022-01-12T17:18:24.190Z",
                "latitudeE7": 377749000,  # 37.7749
                "longitudeE7": -1224194000,  # -122.4194
                "accuracy": 20,
                "source": "WIFI",
                "deviceTag": 1234567890,
                "platformType": "ANDROID"
            },
            {
                "timestamp": "2022-01-12T17:23:24.000Z",
                "latitudeE7": 377750000,  # 37.7750
                "longitudeE7": -1224195000,  # -122.4195
                "accuracy": 15,
                "source": "GPS",
                "altitude": 150
            },
            {
                "timestamp": "2022-01-12T17:28:24.500Z",
                "latitudeE7": 377751000,  # 37.7751
                "longitudeE7": -1224196000,  # -122.4196
                "accuracy": 25
            }
        ]
    }


def test_parse_takeout_timestamp() -> None:
    """Test parsing Google Takeout timestamp strings."""
    # Test standard ISO 8601 format with Z suffix
    timestamp_str = "2022-01-12T17:18:24.190Z"
    dt = parse_takeout_timestamp(timestamp_str)
    assert dt == datetime(2022, 1, 12, 17, 18, 24, 190000, tzinfo=timezone.utc)

    # Test timestamp without milliseconds
    timestamp_str = "2022-01-12T17:18:24Z"
    dt = parse_takeout_timestamp(timestamp_str)
    assert dt == datetime(2022, 1, 12, 17, 18, 24, 0, tzinfo=timezone.utc)

    # Test invalid timestamp
    with pytest.raises(ValueError, match="Failed to parse timestamp"):
        parse_takeout_timestamp("invalid-timestamp")


def test_convert_e7_coordinate() -> None:
    """Test converting Google's E7 coordinate format."""
    # Test latitude conversion (San Francisco)
    lat_e7 = 377749000  # 37.7749 degrees
    lat_decimal = convert_e7_coordinate(lat_e7)
    assert abs(lat_decimal - 37.7749) < 1e-7

    # Test longitude conversion (San Francisco)
    lon_e7 = -1224194000  # -122.4194 degrees
    lon_decimal = convert_e7_coordinate(lon_e7)
    assert abs(lon_decimal - (-122.4194)) < 1e-7

    # Test zero coordinate
    assert convert_e7_coordinate(0) == 0.0


def test_generate_uuid_from_components() -> None:
    """Test UUID generation from components."""
    timestamp1 = datetime(2023, 1, 1, 12, 0, 0, tzinfo=timezone.utc)
    timestamp2 = datetime(2023, 1, 1, 12, 5, 0, tzinfo=timezone.utc)

    # Same inputs should generate the same UUID
    uuid1 = generate_uuid_from_components(
        timestamp1,
        37.7749,
        -122.4194,
        timestamp2,
        37.7750,
        -122.4195,
        DataType.GOOGLE_TAKEOUT_LOCATION,
    )
    uuid2 = generate_uuid_from_components(
        timestamp1,
        37.7749,
        -122.4194,
        timestamp2,
        37.7750,
        -122.4195,
        DataType.GOOGLE_TAKEOUT_LOCATION,
    )

    assert uuid1 == uuid2
    assert len(uuid1) == 16  # Should be 16 bytes

    # Different inputs should generate different UUIDs
    uuid3 = generate_uuid_from_components(
        timestamp1,
        37.7749,
        -122.4194,
        timestamp2,
        37.7751,  # Different latitude
        -122.4195,
        DataType.GOOGLE_TAKEOUT_LOCATION,
    )

    assert uuid1 != uuid3


def test_parse_location_record() -> None:
    """Test parsing individual location records."""
    # Valid record
    record = {
        "timestamp": "2022-01-12T17:18:24.190Z",
        "latitudeE7": 377749000,
        "longitudeE7": -1224194000,
        "accuracy": 20,
    }

    parsed = parse_location_record(record, "Records.json")
    assert parsed is not None
    assert parsed["DATA_TYPE"] == DataType.GOOGLE_TAKEOUT_LOCATION
    assert parsed["SOURCE_FILE"] == "Records.json"
    assert parsed["GPS_LATITUDE"] == 37.7749
    assert parsed["GPS_LONGITUDE"] == -122.4194

    # Record missing timestamp
    record_no_timestamp = {
        "latitudeE7": 377749000,
        "longitudeE7": -1224194000,
    }
    assert parse_location_record(record_no_timestamp, "Records.json") is None

    # Record missing coordinates
    record_no_coords = {
        "timestamp": "2022-01-12T17:18:24.190Z",
    }
    assert parse_location_record(record_no_coords, "Records.json") is None

    # Record with invalid latitude
    record_invalid_lat = {
        "timestamp": "2022-01-12T17:18:24.190Z",
        "latitudeE7": 1000000000,  # 100 degrees (invalid)
        "longitudeE7": -1224194000,
    }
    assert parse_location_record(record_invalid_lat, "Records.json") is None

    # Record with invalid longitude
    record_invalid_lon = {
        "timestamp": "2022-01-12T17:18:24.190Z",
        "latitudeE7": 377749000,
        "longitudeE7": -2000000000,  # -200 degrees (invalid)
    }
    assert parse_location_record(record_invalid_lon, "Records.json") is None


def test_load_google_takeout_records_to_df_nonexistent_file() -> None:
    """Test loading a non-existent Records.json file."""
    nonexistent_file = Path("/nonexistent/Records.json")

    with pytest.raises(FileNotFoundError):
        load_google_takeout_records_to_df(nonexistent_file)


def test_load_google_takeout_records_to_df_valid_file(
    tmp_path: Path, snapshot: SnapshotAssertion
) -> None:
    """Test loading a valid Google Takeout Records.json file."""
    records_data = create_sample_takeout_data()

    records_file = tmp_path / "Records.json"
    with open(records_file, "w", encoding="utf-8") as f:
        json.dump(records_data, f)

    df = load_google_takeout_records_to_df(records_file)
    assert list(zip(df.columns, df.dtypes)) == snapshot
    assert df.to_dict(orient="records") == snapshot


def test_load_google_takeout_records_to_df_empty_locations(
    tmp_path: Path, snapshot: SnapshotAssertion
) -> None:
    """Test loading a Records.json file with empty locations array."""
    records_data = {"locations": []}

    records_file = tmp_path / "empty_Records.json"
    with open(records_file, "w", encoding="utf-8") as f:
        json.dump(records_data, f)

    df = load_google_takeout_records_to_df(records_file)
    assert list(zip(df.columns, df.dtypes)) == snapshot
    assert df.to_dict(orient="records") == snapshot


def test_load_google_takeout_records_to_df_invalid_json(tmp_path: Path) -> None:
    """Test loading an invalid JSON file."""
    invalid_file = tmp_path / "invalid.json"
    invalid_file.write_text("This is not valid JSON")

    with pytest.raises(ValueError, match="Failed to parse JSON file"):
        load_google_takeout_records_to_df(invalid_file)


def test_load_google_takeout_records_to_df_missing_locations_array(tmp_path: Path) -> None:
    """Test loading a JSON file without locations array."""
    invalid_data = {"some_other_key": "value"}

    invalid_file = tmp_path / "no_locations.json"
    with open(invalid_file, "w", encoding="utf-8") as f:
        json.dump(invalid_data, f)

    with pytest.raises(ValueError, match="missing 'locations' array"):
        load_google_takeout_records_to_df(invalid_file)


def test_load_google_takeout_records_to_df_locations_not_array(tmp_path: Path) -> None:
    """Test loading a JSON file where locations is not an array."""
    invalid_data = {"locations": "not an array"}

    invalid_file = tmp_path / "locations_not_array.json"
    with open(invalid_file, "w", encoding="utf-8") as f:
        json.dump(invalid_data, f)

    with pytest.raises(ValueError, match="'locations' is not an array"):
        load_google_takeout_records_to_df(invalid_file)


def test_load_google_takeout_records_to_df_mixed_valid_invalid_records(
    tmp_path: Path, snapshot: SnapshotAssertion
) -> None:
    """Test loading a Records.json file with mix of valid and invalid records."""
    records_data = {
        "locations": [
            # Valid record
            {
                "timestamp": "2022-01-12T17:18:24.190Z",
                "latitudeE7": 377749000,
                "longitudeE7": -1224194000,
                "accuracy": 20
            },
            # Invalid record - missing timestamp
            {
                "latitudeE7": 377750000,
                "longitudeE7": -1224195000,
            },
            # Invalid record - missing coordinates
            {
                "timestamp": "2022-01-12T17:23:24.000Z",
                "accuracy": 15
            },
            # Valid record
            {
                "timestamp": "2022-01-12T17:28:24.500Z",
                "latitudeE7": 377751000,
                "longitudeE7": -1224196000,
                "accuracy": 25
            },
            # Invalid record - non-dict
            "not a dict",
            # Invalid record - bad coordinates
            {
                "timestamp": "2022-01-12T17:33:24.000Z",
                "latitudeE7": 1000000000,  # Invalid latitude (100 degrees)
                "longitudeE7": -1224196000,
            }
        ]
    }

    records_file = tmp_path / "mixed_Records.json"
    with open(records_file, "w", encoding="utf-8") as f:
        json.dump(records_data, f)

    df = load_google_takeout_records_to_df(records_file)
    # Should only load the 2 valid records
    assert len(df) == 2
    assert list(zip(df.columns, df.dtypes)) == snapshot
    assert df.to_dict(orient="records") == snapshot


def test_load_google_takeout_records_to_df_all_invalid_records(
    tmp_path: Path, snapshot: SnapshotAssertion
) -> None:
    """Test loading a Records.json file where all records are invalid."""
    records_data = {
        "locations": [
            # Missing timestamp
            {
                "latitudeE7": 377749000,
                "longitudeE7": -1224194000,
            },
            # Missing coordinates
            {
                "timestamp": "2022-01-12T17:18:24.190Z",
                "accuracy": 20
            },
            # Invalid coordinates
            {
                "timestamp": "2022-01-12T17:23:24.000Z",
                "latitudeE7": 1000000000,  # 100 degrees (invalid)
                "longitudeE7": -2000000000,  # -200 degrees (invalid)
            }
        ]
    }

    records_file = tmp_path / "all_invalid_Records.json"
    with open(records_file, "w", encoding="utf-8") as f:
        json.dump(records_data, f)

    df = load_google_takeout_records_to_df(records_file)
    assert len(df) == 0
    assert list(zip(df.columns, df.dtypes)) == snapshot
    assert df.to_dict(orient="records") == snapshot