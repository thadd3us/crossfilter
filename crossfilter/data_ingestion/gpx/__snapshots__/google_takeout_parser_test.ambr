# serializer version: 1
# name: test_load_google_takeout_records_to_df_all_invalid_records
  list([
    tuple(
      <SchemaColumns.UUID: 'UUID'>,
      dtype('O'),
    ),
    tuple(
      <SchemaColumns.DATA_TYPE: 'DATA_TYPE'>,
      dtype('O'),
    ),
    tuple(
      <SchemaColumns.NAME: 'NAME'>,
      dtype('O'),
    ),
    tuple(
      <SchemaColumns.SOURCE_FILE: 'SOURCE_FILE'>,
      dtype('O'),
    ),
    tuple(
      <SchemaColumns.TIMESTAMP_MAYBE_TIMEZONE_AWARE: 'TIMESTAMP_MAYBE_TIMEZONE_AWARE'>,
      dtype('O'),
    ),
    tuple(
      <SchemaColumns.TIMESTAMP_UTC: 'TIMESTAMP_UTC'>,
      dtype('O'),
    ),
    tuple(
      <SchemaColumns.GPS_LATITUDE: 'GPS_LATITUDE'>,
      dtype('O'),
    ),
    tuple(
      <SchemaColumns.GPS_LONGITUDE: 'GPS_LONGITUDE'>,
      dtype('O'),
    ),
  ])
# ---
# name: test_load_google_takeout_records_to_df_all_invalid_records.1
  list([
  ])
# ---
# name: test_load_google_takeout_records_to_df_empty_locations
  list([
    tuple(
      <SchemaColumns.UUID: 'UUID'>,
      dtype('O'),
    ),
    tuple(
      <SchemaColumns.DATA_TYPE: 'DATA_TYPE'>,
      dtype('O'),
    ),
    tuple(
      <SchemaColumns.NAME: 'NAME'>,
      dtype('O'),
    ),
    tuple(
      <SchemaColumns.SOURCE_FILE: 'SOURCE_FILE'>,
      dtype('O'),
    ),
    tuple(
      <SchemaColumns.TIMESTAMP_MAYBE_TIMEZONE_AWARE: 'TIMESTAMP_MAYBE_TIMEZONE_AWARE'>,
      dtype('O'),
    ),
    tuple(
      <SchemaColumns.TIMESTAMP_UTC: 'TIMESTAMP_UTC'>,
      dtype('O'),
    ),
    tuple(
      <SchemaColumns.GPS_LATITUDE: 'GPS_LATITUDE'>,
      dtype('O'),
    ),
    tuple(
      <SchemaColumns.GPS_LONGITUDE: 'GPS_LONGITUDE'>,
      dtype('O'),
    ),
  ])
# ---
# name: test_load_google_takeout_records_to_df_empty_locations.1
  list([
  ])
# ---
# name: test_load_google_takeout_records_to_df_mixed_valid_invalid_records
  list([
    tuple(
      <SchemaColumns.DATA_TYPE: 'DATA_TYPE'>,
      dtype('O'),
    ),
    tuple(
      <SchemaColumns.SOURCE_FILE: 'SOURCE_FILE'>,
      dtype('O'),
    ),
    tuple(
      <SchemaColumns.TIMESTAMP_MAYBE_TIMEZONE_AWARE: 'TIMESTAMP_MAYBE_TIMEZONE_AWARE'>,
      dtype('O'),
    ),
    tuple(
      <SchemaColumns.TIMESTAMP_UTC: 'TIMESTAMP_UTC'>,
      datetime64[ns, UTC],
    ),
    tuple(
      <SchemaColumns.GPS_LATITUDE: 'GPS_LATITUDE'>,
      dtype('float64'),
    ),
    tuple(
      <SchemaColumns.GPS_LONGITUDE: 'GPS_LONGITUDE'>,
      dtype('float64'),
    ),
    tuple(
      <SchemaColumns.UUID: 'UUID'>,
      dtype('O'),
    ),
    tuple(
      'BUCKETED_H3_L0',
      dtype('O'),
    ),
    tuple(
      'BUCKETED_H3_L1',
      dtype('O'),
    ),
    tuple(
      'BUCKETED_H3_L2',
      dtype('O'),
    ),
    tuple(
      'BUCKETED_H3_L3',
      dtype('O'),
    ),
    tuple(
      'BUCKETED_H3_L4',
      dtype('O'),
    ),
    tuple(
      'BUCKETED_H3_L5',
      dtype('O'),
    ),
    tuple(
      'BUCKETED_H3_L6',
      dtype('O'),
    ),
    tuple(
      'BUCKETED_H3_L7',
      dtype('O'),
    ),
    tuple(
      'BUCKETED_H3_L8',
      dtype('O'),
    ),
    tuple(
      'BUCKETED_H3_L9',
      dtype('O'),
    ),
    tuple(
      'BUCKETED_H3_L10',
      dtype('O'),
    ),
    tuple(
      'BUCKETED_H3_L11',
      dtype('O'),
    ),
    tuple(
      'BUCKETED_H3_L12',
      dtype('O'),
    ),
    tuple(
      'BUCKETED_H3_L13',
      dtype('O'),
    ),
    tuple(
      'BUCKETED_H3_L14',
      dtype('O'),
    ),
    tuple(
      'BUCKETED_H3_L15',
      dtype('O'),
    ),
  ])
# ---
# name: test_load_google_takeout_records_to_df_mixed_valid_invalid_records.1
  list([
    dict({
      'BUCKETED_H3_L0': '8029fffffffffff',
      'BUCKETED_H3_L1': '81283ffffffffff',
      'BUCKETED_H3_L10': '8a283082800ffff',
      'BUCKETED_H3_L11': '8b283082800bfff',
      'BUCKETED_H3_L12': '8c283082800b3ff',
      'BUCKETED_H3_L13': '8d283082800b3bf',
      'BUCKETED_H3_L14': '8e283082800b397',
      'BUCKETED_H3_L15': '8f283082800b390',
      'BUCKETED_H3_L2': '822837fffffffff',
      'BUCKETED_H3_L3': '832830fffffffff',
      'BUCKETED_H3_L4': '8428309ffffffff',
      'BUCKETED_H3_L5': '85283083fffffff',
      'BUCKETED_H3_L6': '86283082fffffff',
      'BUCKETED_H3_L7': '872830828ffffff',
      'BUCKETED_H3_L8': '8828308281fffff',
      'BUCKETED_H3_L9': '89283082803ffff',
      <SchemaColumns.DATA_TYPE: 'DATA_TYPE'>: <DataType.GOOGLE_TAKEOUT_LOCATION: 'GOOGLE_TAKEOUT_LOCATION'>,
      <SchemaColumns.GPS_LATITUDE: 'GPS_LATITUDE'>: 37.7749,
      <SchemaColumns.GPS_LONGITUDE: 'GPS_LONGITUDE'>: -122.4194,
      <SchemaColumns.SOURCE_FILE: 'SOURCE_FILE'>: 'mixed_Records.json',
      <SchemaColumns.TIMESTAMP_MAYBE_TIMEZONE_AWARE: 'TIMESTAMP_MAYBE_TIMEZONE_AWARE'>: '2022-01-12T17:18:24.190Z',
      <SchemaColumns.TIMESTAMP_UTC: 'TIMESTAMP_UTC'>: Timestamp('2022-01-12 17:18:24.190000+0000', tz='UTC'),
      <SchemaColumns.UUID: 'UUID'>: b"K9\x1f\x83\x93\x1ani\x15\xa0v\r\xac\xab'\x03",
    }),
    dict({
      'BUCKETED_H3_L0': '8029fffffffffff',
      'BUCKETED_H3_L1': '81283ffffffffff',
      'BUCKETED_H3_L10': '8a2830828057fff',
      'BUCKETED_H3_L11': '8b2830828056fff',
      'BUCKETED_H3_L12': '8c28308280565ff',
      'BUCKETED_H3_L13': '8d28308280564bf',
      'BUCKETED_H3_L14': '8e283082805648f',
      'BUCKETED_H3_L15': '8f2830828056488',
      'BUCKETED_H3_L2': '822837fffffffff',
      'BUCKETED_H3_L3': '832830fffffffff',
      'BUCKETED_H3_L4': '8428309ffffffff',
      'BUCKETED_H3_L5': '85283083fffffff',
      'BUCKETED_H3_L6': '86283082fffffff',
      'BUCKETED_H3_L7': '872830828ffffff',
      'BUCKETED_H3_L8': '8828308281fffff',
      'BUCKETED_H3_L9': '89283082807ffff',
      <SchemaColumns.DATA_TYPE: 'DATA_TYPE'>: <DataType.GOOGLE_TAKEOUT_LOCATION: 'GOOGLE_TAKEOUT_LOCATION'>,
      <SchemaColumns.GPS_LATITUDE: 'GPS_LATITUDE'>: 37.7751,
      <SchemaColumns.GPS_LONGITUDE: 'GPS_LONGITUDE'>: -122.4196,
      <SchemaColumns.SOURCE_FILE: 'SOURCE_FILE'>: 'mixed_Records.json',
      <SchemaColumns.TIMESTAMP_MAYBE_TIMEZONE_AWARE: 'TIMESTAMP_MAYBE_TIMEZONE_AWARE'>: '2022-01-12T17:28:24.500Z',
      <SchemaColumns.TIMESTAMP_UTC: 'TIMESTAMP_UTC'>: Timestamp('2022-01-12 17:28:24.500000+0000', tz='UTC'),
      <SchemaColumns.UUID: 'UUID'>: b'\xc2\xd2\xf4\xb0\x92`;"\xb6\xd8\xb35\x0e~@\x1c',
    }),
  ])
# ---
# name: test_load_google_takeout_records_to_df_valid_file
  list([
    tuple(
      <SchemaColumns.DATA_TYPE: 'DATA_TYPE'>,
      dtype('O'),
    ),
    tuple(
      <SchemaColumns.SOURCE_FILE: 'SOURCE_FILE'>,
      dtype('O'),
    ),
    tuple(
      <SchemaColumns.TIMESTAMP_MAYBE_TIMEZONE_AWARE: 'TIMESTAMP_MAYBE_TIMEZONE_AWARE'>,
      dtype('O'),
    ),
    tuple(
      <SchemaColumns.TIMESTAMP_UTC: 'TIMESTAMP_UTC'>,
      datetime64[ns, UTC],
    ),
    tuple(
      <SchemaColumns.GPS_LATITUDE: 'GPS_LATITUDE'>,
      dtype('float64'),
    ),
    tuple(
      <SchemaColumns.GPS_LONGITUDE: 'GPS_LONGITUDE'>,
      dtype('float64'),
    ),
    tuple(
      <SchemaColumns.UUID: 'UUID'>,
      dtype('O'),
    ),
    tuple(
      'BUCKETED_H3_L0',
      dtype('O'),
    ),
    tuple(
      'BUCKETED_H3_L1',
      dtype('O'),
    ),
    tuple(
      'BUCKETED_H3_L2',
      dtype('O'),
    ),
    tuple(
      'BUCKETED_H3_L3',
      dtype('O'),
    ),
    tuple(
      'BUCKETED_H3_L4',
      dtype('O'),
    ),
    tuple(
      'BUCKETED_H3_L5',
      dtype('O'),
    ),
    tuple(
      'BUCKETED_H3_L6',
      dtype('O'),
    ),
    tuple(
      'BUCKETED_H3_L7',
      dtype('O'),
    ),
    tuple(
      'BUCKETED_H3_L8',
      dtype('O'),
    ),
    tuple(
      'BUCKETED_H3_L9',
      dtype('O'),
    ),
    tuple(
      'BUCKETED_H3_L10',
      dtype('O'),
    ),
    tuple(
      'BUCKETED_H3_L11',
      dtype('O'),
    ),
    tuple(
      'BUCKETED_H3_L12',
      dtype('O'),
    ),
    tuple(
      'BUCKETED_H3_L13',
      dtype('O'),
    ),
    tuple(
      'BUCKETED_H3_L14',
      dtype('O'),
    ),
    tuple(
      'BUCKETED_H3_L15',
      dtype('O'),
    ),
  ])
# ---
# name: test_load_google_takeout_records_to_df_valid_file.1
  list([
    dict({
      'BUCKETED_H3_L0': '8029fffffffffff',
      'BUCKETED_H3_L1': '81283ffffffffff',
      'BUCKETED_H3_L10': '8a283082800ffff',
      'BUCKETED_H3_L11': '8b283082800bfff',
      'BUCKETED_H3_L12': '8c283082800b3ff',
      'BUCKETED_H3_L13': '8d283082800b3bf',
      'BUCKETED_H3_L14': '8e283082800b397',
      'BUCKETED_H3_L15': '8f283082800b390',
      'BUCKETED_H3_L2': '822837fffffffff',
      'BUCKETED_H3_L3': '832830fffffffff',
      'BUCKETED_H3_L4': '8428309ffffffff',
      'BUCKETED_H3_L5': '85283083fffffff',
      'BUCKETED_H3_L6': '86283082fffffff',
      'BUCKETED_H3_L7': '872830828ffffff',
      'BUCKETED_H3_L8': '8828308281fffff',
      'BUCKETED_H3_L9': '89283082803ffff',
      <SchemaColumns.DATA_TYPE: 'DATA_TYPE'>: <DataType.GOOGLE_TAKEOUT_LOCATION: 'GOOGLE_TAKEOUT_LOCATION'>,
      <SchemaColumns.GPS_LATITUDE: 'GPS_LATITUDE'>: 37.7749,
      <SchemaColumns.GPS_LONGITUDE: 'GPS_LONGITUDE'>: -122.4194,
      <SchemaColumns.SOURCE_FILE: 'SOURCE_FILE'>: 'Records.json',
      <SchemaColumns.TIMESTAMP_MAYBE_TIMEZONE_AWARE: 'TIMESTAMP_MAYBE_TIMEZONE_AWARE'>: '2022-01-12T17:18:24.190Z',
      <SchemaColumns.TIMESTAMP_UTC: 'TIMESTAMP_UTC'>: Timestamp('2022-01-12 17:18:24.190000+0000', tz='UTC'),
      <SchemaColumns.UUID: 'UUID'>: b"K9\x1f\x83\x93\x1ani\x15\xa0v\r\xac\xab'\x03",
    }),
    dict({
      'BUCKETED_H3_L0': '8029fffffffffff',
      'BUCKETED_H3_L1': '81283ffffffffff',
      'BUCKETED_H3_L10': '8a283082800ffff',
      'BUCKETED_H3_L11': '8b283082800bfff',
      'BUCKETED_H3_L12': '8c283082800b3ff',
      'BUCKETED_H3_L13': '8d283082800b2ff',
      'BUCKETED_H3_L14': '8e283082800b2c7',
      'BUCKETED_H3_L15': '8f283082800b2c3',
      'BUCKETED_H3_L2': '822837fffffffff',
      'BUCKETED_H3_L3': '832830fffffffff',
      'BUCKETED_H3_L4': '8428309ffffffff',
      'BUCKETED_H3_L5': '85283083fffffff',
      'BUCKETED_H3_L6': '86283082fffffff',
      'BUCKETED_H3_L7': '872830828ffffff',
      'BUCKETED_H3_L8': '8828308281fffff',
      'BUCKETED_H3_L9': '89283082803ffff',
      <SchemaColumns.DATA_TYPE: 'DATA_TYPE'>: <DataType.GOOGLE_TAKEOUT_LOCATION: 'GOOGLE_TAKEOUT_LOCATION'>,
      <SchemaColumns.GPS_LATITUDE: 'GPS_LATITUDE'>: 37.775,
      <SchemaColumns.GPS_LONGITUDE: 'GPS_LONGITUDE'>: -122.4195,
      <SchemaColumns.SOURCE_FILE: 'SOURCE_FILE'>: 'Records.json',
      <SchemaColumns.TIMESTAMP_MAYBE_TIMEZONE_AWARE: 'TIMESTAMP_MAYBE_TIMEZONE_AWARE'>: '2022-01-12T17:23:24.000Z',
      <SchemaColumns.TIMESTAMP_UTC: 'TIMESTAMP_UTC'>: Timestamp('2022-01-12 17:23:24+0000', tz='UTC'),
      <SchemaColumns.UUID: 'UUID'>: b'\xd9\xad\x02D$\xb2\x12\xa8\x03\x86\x0e\x08\xa8Rw\xa7',
    }),
    dict({
      'BUCKETED_H3_L0': '8029fffffffffff',
      'BUCKETED_H3_L1': '81283ffffffffff',
      'BUCKETED_H3_L10': '8a2830828057fff',
      'BUCKETED_H3_L11': '8b2830828056fff',
      'BUCKETED_H3_L12': '8c28308280565ff',
      'BUCKETED_H3_L13': '8d28308280564bf',
      'BUCKETED_H3_L14': '8e283082805648f',
      'BUCKETED_H3_L15': '8f2830828056488',
      'BUCKETED_H3_L2': '822837fffffffff',
      'BUCKETED_H3_L3': '832830fffffffff',
      'BUCKETED_H3_L4': '8428309ffffffff',
      'BUCKETED_H3_L5': '85283083fffffff',
      'BUCKETED_H3_L6': '86283082fffffff',
      'BUCKETED_H3_L7': '872830828ffffff',
      'BUCKETED_H3_L8': '8828308281fffff',
      'BUCKETED_H3_L9': '89283082807ffff',
      <SchemaColumns.DATA_TYPE: 'DATA_TYPE'>: <DataType.GOOGLE_TAKEOUT_LOCATION: 'GOOGLE_TAKEOUT_LOCATION'>,
      <SchemaColumns.GPS_LATITUDE: 'GPS_LATITUDE'>: 37.7751,
      <SchemaColumns.GPS_LONGITUDE: 'GPS_LONGITUDE'>: -122.4196,
      <SchemaColumns.SOURCE_FILE: 'SOURCE_FILE'>: 'Records.json',
      <SchemaColumns.TIMESTAMP_MAYBE_TIMEZONE_AWARE: 'TIMESTAMP_MAYBE_TIMEZONE_AWARE'>: '2022-01-12T17:28:24.500Z',
      <SchemaColumns.TIMESTAMP_UTC: 'TIMESTAMP_UTC'>: Timestamp('2022-01-12 17:28:24.500000+0000', tz='UTC'),
      <SchemaColumns.UUID: 'UUID'>: b'\xc2\xd2\xf4\xb0\x92`;"\xb6\xd8\xb35\x0e~@\x1c',
    }),
  ])
# ---
